Mathlib/NumberTheory/MahlerMeasure.lean

```diff
@@ -0,0 +1,113 @@
+/-
+Copyright (c) 2025 Fabrizio Barroero. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fabrizio Barroero
+-/
+module
+
+public import Mathlib.Algebra.Polynomial.OfFn
+public import Mathlib.Analysis.Polynomial.MahlerMeasure
+public import Mathlib.Data.Pi.Interval
+
+/-!
+# Mahler measure of integer polynomials
+
+The main purpose of this file is to prove Northcott's Theorem for the Mahler measure.
+
+## Main results
+
+- `Polynomial.card_mahlerMeasure_le_prod`: an upper bound on the number of integer polynomials
+  of degree at most `n` and Mahler measure at most `B`, also known as Northcott's Theorem.
+-/
+
+namespace Polynomial
+
+open Int
+
+private lemma degree_map_int_castRingHom (p : ℤ[X]) :
+    (map (castRingHom ℂ) p).natDegree = p.natDegree := by
+  rcases eq_or_ne p 0 with rfl | hp
+  · simp
+  · simp [hp]
+
+lemma one_le_mahlerMeasure_of_ne_zero {p : ℤ[X]} (hp : p ≠ 0) :
+    1 ≤ (p.map (castRingHom ℂ)).mahlerMeasure := by
+  apply le_trans _ (p.map (castRingHom ℂ)).leading_coeff_le_mahlerMeasure
+  rw [leadingCoeff, p.degree_map_int_castRingHom]
+  simp only [coeff_map, coeff_natDegree, eq_intCast, Complex.norm_intCast]
+  exact_mod_cast one_le_abs <| leadingCoeff_ne_zero.mpr hp
+
+section Northcott
+
+variable {n : ℕ} {B₁ B₂ : Fin (n + 1) → ℝ}
+
+local notation3 "BoxPoly" =>
+  {p : ℤ[X] | p.natDegree ≤ n ∧ ∀ i : Fin (n + 1), B₁ i ≤ p.coeff i ∧ p.coeff i ≤ B₂ i}
+
+open Finset in
+theorem card_eq_of_natDegree_le_of_coeff_le (h_B : ∀ i, ⌈B₁ i⌉ ≤ ⌊B₂ i⌋) :
+    Nat.card BoxPoly = ∏ i : Fin (n + 1), (⌊B₂ i⌋ - ⌈B₁ i⌉ + 1).toNat  := by
```

Comment:
I would think that `Set.ncard` would be more natural here (avoids casting the set to a type).
On the other hand, the API for `Set.ncard` could use some additions (e.g., a lemma that says that two sets with an equivalence between them have the same cardinality).